File: server/app/services/scoring/reliability_adjuster.rb
# frozen_string_literal: true

module Scoring
  # Evaluates the reliability of an assessment's responses.
  #
  # Four checks are performed:
  #
  # 1. Split-half consistency: Correlates odd-positioned vs even-positioned item
  #    scores within each domain. Low correlation suggests random or inattentive
  #    responding. Uses Spearman-Brown corrected Pearson r.
  #
  # 2. Speed anomalies: Flags when more than 50% of responses have a
  #    response_time_ms under 500ms, suggesting the respondent is clicking
  #    through without reading.
  #
  # 3. Non-response rate: Fraction of questions left unanswered (value: nil).
  #
  # 4. Extreme response rate: Fraction of answered questions where the value
  #    is 1 or 5, suggesting an acquiescence or extreme response bias.
  #
  # Returns:
  #   {
  #     reliability_coefficient: 0.82,
  #     consistency_index: 0.75,
  #     speed_flag: false,
  #     non_response_rate: 0.05,
  #     extreme_response_rate: 0.30,
  #     flags: ["high_extreme_response_rate"]
  #   }
  #
  class ReliabilityAdjuster
    SPEED_THRESHOLD_MS = 500
    SPEED_ANOMALY_RATIO = 0.5
    NON_RESPONSE_WARNING = 0.2
    EXTREME_RESPONSE_WARNING = 0.6

    attr_reader :assessment

    def initialize(assessment)
      @assessment = assessment
    end

    def call
      flags = []

      flags << "low_split_half_consistency" if consistency_index < 0.5
      flags << "speed_anomaly"              if speed_flag?
      flags << "high_non_response_rate"     if non_response_rate > NON_RESPONSE_WARNING
      flags << "high_extreme_response_rate" if extreme_response_rate > EXTREME_RESPONSE_WARNING

      {
        reliability_coefficient: reliability_coefficient,
        consistency_index: consistency_index,
        speed_flag: speed_flag?,
        non_response_rate: non_response_rate,
        extreme_response_rate: extreme_response_rate,
        flags: flags
      }
    end

    private

    # ------------------------------------------------------------------
    # Composite reliability: weighted average of sub-indicators, each in 0-1.
    # ------------------------------------------------------------------
    def reliability_coefficient
      @reliability_coefficient ||= begin
        weights = { consistency: 0.4, speed: 0.2, non_response: 0.2, extreme: 0.2 }

        score = 0.0
        score += weights[:consistency] * consistency_index
        score += weights[:speed]       * (speed_flag? ? 0.0 : 1.0)
        score += weights[:non_response] * (1.0 - non_response_rate)
        score += weights[:extreme]      * (1.0 - extreme_response_rate)

        score.round(3)
      end
    end

    # ------------------------------------------------------------------
    # Split-half consistency across all domains (averaged).
    # ------------------------------------------------------------------
    def consistency_index
      @consistency_index ||= begin
        domain_consistencies = Question::DOMAINS.filter_map do |domain|
          split_half_for(domain)
        end

        return 0.0 if domain_consistencies.empty?

        (domain_consistencies.sum / domain_consistencies.size).round(3)
      end
    end

    # Pearson r between odd- and even-positioned item scores within a domain,
    # corrected via Spearman-Brown prophecy formula: r_sb = 2r / (1 + r).
    def split_half_for(domain)
      domain_resp = responses_with_questions.select do |r|
        r.question.domain == domain && !r.skipped?
      end

      odd_values  = []
      even_values = []

      domain_resp.sort_by { |r| r.question.position }.each do |r|
        val = r.question.negative? ? (6 - r.value) : r.value
        if r.question.position.odd?
          odd_values << val
        else
          even_values << val
        end
      end

      # Need at least 2 items per half to compute a meaningful correlation.
      return nil if odd_values.size < 2 || even_values.size < 2

      # Truncate to equal length for pairing.
      min_len = [odd_values.size, even_values.size].min
      r = pearson_r(odd_values.first(min_len), even_values.first(min_len))

      return 0.0 if r.nil?

      # Spearman-Brown correction.
      corrected = (2.0 * r) / (1.0 + r)
      corrected.clamp(0.0, 1.0)
    end

    # ------------------------------------------------------------------
    # Speed anomaly detection.
    # ------------------------------------------------------------------
    def speed_flag?
      return @speed_flag if defined?(@speed_flag)

      @speed_flag = begin
        timed = responses_with_questions.select { |r| r.response_time_ms.present? }
        return false if timed.empty?

        fast_count = timed.count { |r| r.response_time_ms < SPEED_THRESHOLD_MS }
        (fast_count.to_f / timed.size) > SPEED_ANOMALY_RATIO
      end
    end

    # ------------------------------------------------------------------
    # Non-response rate.
    # ------------------------------------------------------------------
    def non_response_rate
      @non_response_rate ||= begin
        total = responses_with_questions.size
        return 0.0 if total.zero?

        skipped = responses_with_questions.count(&:skipped?)
        (skipped.to_f / total).round(3)
      end
    end

    # ------------------------------------------------------------------
    # Extreme response rate (1s and 5s among answered items).
    # ------------------------------------------------------------------
    def extreme_response_rate
      @extreme_response_rate ||= begin
        answered = responses_with_questions.reject(&:skipped?)
        return 0.0 if answered.empty?

        extremes = answered.count(&:extreme?)
        (extremes.to_f / answered.size).round(3)
      end
    end

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def responses_with_questions
      @responses_with_questions ||= assessment.responses.includes(:question).to_a
    end

    # Standard Pearson correlation coefficient for two equal-length arrays.
    def pearson_r(xs, ys)
      n = xs.size
      return nil if n < 2

      mean_x = xs.sum.to_f / n
      mean_y = ys.sum.to_f / n

      numerator   = 0.0
      denom_x_sq  = 0.0
      denom_y_sq  = 0.0

      n.times do |i|
        dx = xs[i] - mean_x
        dy = ys[i] - mean_y
        numerator  += dx * dy
        denom_x_sq += dx * dx
        denom_y_sq += dy * dy
      end

      denominator = Math.sqrt(denom_x_sq * denom_y_sq)
      return nil if denominator.zero?

      numerator / denominator
    end
  end
end

File: apps/workers/src/services/scoring/reliabilityAdjuster.ts
/**
 * src/services/scoring/reliabilityAdjuster.ts — GREEN phase
 * Rails: server/app/services/scoring/reliability_adjuster.rb
 *
 * adjustReliability(responses) → ReliabilityResult
 * reliability_coefficient = 0.4*consistency + 0.2*(speed?0:1) + 0.2*(1-non_response) + 0.2*(1-extreme)
 * Pearson r: split-half correlation
 * Spearman-Brown: 2r/(1+r), clamped to [0,1]
 * Flags: "low_split_half_consistency", "high_extreme_response_rate", "speed_anomaly", "high_non_response_rate"
 */

export interface ReliabilityResult {
  reliability_coefficient: number;
  consistency_index: number;
  speed_flag: boolean;
  non_response_rate: number;
  extreme_response_rate: number;
  flags: string[];
}

export interface ResponseForReliability {
  value: number | null;
  response_time_ms: number | null;
}

/**
 * Pearson correlation coefficient between two arrays.
 * Returns null if arrays have < 2 elements or zero variance in either array.
 */
export function pearsonR(a: number[], b: number[]): number | null {
  if (a.length < 2 || b.length < 2 || a.length !== b.length) return null;

  const n = a.length;
  const meanA = a.reduce((s, v) => s + v, 0) / n;
  const meanB = b.reduce((s, v) => s + v, 0) / n;

  let cov = 0;
  let varA = 0;
  let varB = 0;

  for (let i = 0; i < n; i++) {
    const da = a[i] - meanA;
    const db = b[i] - meanB;
    cov += da * db;
    varA += da * da;
    varB += db * db;
  }

  if (varA === 0 || varB === 0) return null;

  return cov / Math.sqrt(varA * varB);
}

/**
 * Spearman-Brown prophecy formula: 2r/(1+r), clamped to [0,1].
 */
function spearmanBrown(r: number): number {
  const sb = (2 * r) / (1 + r);
  return Math.max(0, Math.min(1, sb));
}

export function adjustReliability(
  responses: ResponseForReliability[]
): ReliabilityResult {
  if (responses.length === 0) {
    return {
      reliability_coefficient: 0.2, // 0.4*0 + 0.2*1 + 0.2*1 + 0.2*1
      consistency_index: 0.0,
      speed_flag: false,
      non_response_rate: 0.0,
      extreme_response_rate: 0.0,
      flags: ["low_split_half_consistency"],
    };
  }

  const total = responses.length;
  const answered = responses.filter((r) => r.value != null);
  const answeredValues = answered.map((r) => r.value as number);

  // Non-response rate
  const non_response_rate = (total - answered.length) / total;

  // Extreme response rate (value 1 or 5)
  const extremeCount = answeredValues.filter((v) => v === 1 || v === 5).length;
  const extreme_response_rate = total > 0 ? extremeCount / total : 0.0;

  // Speed flag: > 50% of responses < 500ms
  const times = responses.map((r) => r.response_time_ms).filter((t): t is number => t != null);
  const fastCount = times.filter((t) => t < 500).length;
  const speed_flag = times.length > 0 && fastCount / times.length > 0.5;

  // Split-half consistency via Pearson r + Spearman-Brown
  let consistency_index = 0.0;
  if (answeredValues.length >= 4) {
    const half = Math.floor(answeredValues.length / 2);
    const firstHalf = answeredValues.slice(0, half);
    const secondHalf = answeredValues.slice(half, half * 2);
    const r = pearsonR(firstHalf, secondHalf);
    if (r != null && r > 0) {
      consistency_index = spearmanBrown(r);
    }
  }

  // Flags
  const flags: string[] = [];
  if (consistency_index < 0.5) {
    flags.push("low_split_half_consistency");
  }
  if (extreme_response_rate >= 0.8) {
    flags.push("high_extreme_response_rate");
  }
  if (speed_flag) {
    flags.push("speed_anomaly");
  }
  if (non_response_rate > 0.2) {
    flags.push("high_non_response_rate");
  }

  // Reliability coefficient (weighted)
  const reliability_coefficient =
    0.4 * consistency_index +
    0.2 * (speed_flag ? 0.0 : 1.0) +
    0.2 * (1.0 - non_response_rate) +
    0.2 * (1.0 - extreme_response_rate);

  return {
    reliability_coefficient: Math.round(reliability_coefficient * 1000) / 1000,
    consistency_index,
    speed_flag,
    non_response_rate,
    extreme_response_rate,
    flags,
  };
}
